### ISSUE TITLE
arcgis-js-api alternate typings 

### ISSUE THREAD

**Issue by TwitchBronBron** [DefinitelyTyped/DefinitelyTyped#15821]
- [x] I tried using the `@types/arcgis-js-api` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @ESRI

This is in relation to the [arcgis-js-api v3](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/arcgis-js-api/v3) definitions. 

## Overview
I have created namespace-style type definitions for v3 of the arcgis-js-api for people who don't use AMD for their whole project. The definitions can be found here: https://github.com/TwitchBronBron/esri-javascript-api-typings. Where in DefinitelyTyped should I put them, and what should I name them since they are just in a different format from the original?

## Details
The arcgis-js-api definitions are written assuming that all users will be using the AMD module format. However, we do not write our projects in AMD. Instead, we use standard typescript namespaces, and sprinkle in AMD whenever we need to use the arcgis-js-api pieces. Here's an example: 

```TypeScript
namespace app {
    class MapManager {

        private map: any;

        createMap(): Promise<any> {
            return new Promise((resolve) => {
                require(['esri/map'], function (Map) {
                    this.map = new Map('map-id');
                    resolve(this.map);
                });
            });
        }
    }
}  
```

This prevents us from using the arcgis-js-api TypeScript definitions found in DefinitelyTyped because we are not all-in with AMD.

For everyone who uses the ESRI JavaScript API in this way, it would be valuable to have the types in a standard namespace-style format. We would need to manually assign all of those types to our objects, but there is still a TON of value in having type information about these once they are wired up. Here is how we would consume those definitions:

```TypeScript
/// <reference path="path/to/arcgis-js-api-namespaced.d.ts" />
namespace app {
    class MapManager {
       
         //Now every consumer of MapManager#map will get type information about the esri map
        public map: esri.Map;

        createMap(): Promise<esri.Map> {
            return new Promise((resolve) => {
                require(['esri/map'], function (Map) {
                    this.map = new Map('map-id');
                    resolve(this.map);
                });
            });
        }
    }
}
```

I have already created the namespace-style versions of the arcgis-js-api type definitions. You can find them here: https://github.com/TwitchBronBron/esri-javascript-api-typings. I think these belong in DefinitelyTyped, but **where in DefinitelyTyped should I put them**, and **what should I name it** since there is already a project with arcgis-js-api in this repository? 

Thanks!

**Comment by TwitchBronBron** [comment]
Correctly referenced author

**Comment by andy-ms** [comment]
In TypeScript, if an `import` statement only pulls in types, it will be elided from the compiled output. So you won't need AMD support at runtime. So I don't think there's a need to put everything in namespaces if you are just referencing them for the types.

**Comment by TwitchBronBron** [comment]
When I tried adding this import statement to one of our files, I got this error: 

![image](https://cloud.githubusercontent.com/assets/2544493/24968293/d2a97ef0-1f7a-11e7-9c31-084180c4c0f5.png)

Once I added the import statement, a bunch of other typescript errors started being thrown too, like this one:
 - src\services\symbolService.ts(48,28): error TS2686: 'angular' refers to a UMD global, but the current file is a module. Consider adding an import instead.

In our TypeScript compilerOptions, we are using module: "none". We use an external build tool to combine all of our typescript files into a single file. 

I'm more than happy to try any other ideas that you have so that we can use the existing type definitions. However, I'm thinking that for projects who don't use a module system, there may not be a way to use the existing definitions without causing errors.

**Comment by andy-ms** [comment]
The UMD issue looks like Microsoft/TypeScript#10178.

It would also be nice to have `import type * as esri from "arcgis-js-api";`, from Microsoft/TypeScript#14853.

Are you combining the *output* into a single file, or are you concatenating TypeScript source code? In either case, your build tool might be able to remove even value imports (https://github.com/Microsoft/TypeScript/issues/10178#issuecomment-285357649).

In either case, you can still use imports if you set `"module"`, i.e.:

**tsconfig.json**

```json
{
    "compilerOptions": {
        "target": "es3",
        "lib": ["es6"],
        "module": "es6",
        "moduleResolution": "node"
    }
}
```

**app.ts**

```ts
import * as Map from "esri/Map";

declare function require(strs: string[], cb: (modules: any[]) => void): void;

namespace app {
    function getMap(): Promise<Map> {
        return new Promise(resolve => {
            require(["esri/map"], Map => {
                resolve(Map);
            });
        });
    }
}
```

**-> app.js**

```
var app;
(function (app) {
    function getMap() {
        return new Promise(function (resolve) {
            require(["esri/map"], function (Map) {
                resolve(Map);
            });
        });
    }
})(app || (app = {}));
```

Although you would now have be be sure to not accidentally import something as a value.

Also, the current declaration seems to define a global `__esri` namespace, although I wouldn't rely on it, since we don't prefer that style and might remove it in the future.

**Comment by TwitchBronBron** [comment]
We run all of our TypeScript files through the typescript compiler, and then concat the output into a single file. 

The problem is, once you have an import statement in your typescript file, it seems that TypeScript now considers that file a module and is its namespace is no longer accessible in the rest of the project. 

Here is quick sample application that showcases how our project is set up. https://github.com/TwitchBronBron/esri-javascript-api-typings-example/

Once you uncomment the first line in MapManager.ts, the main.ts file no longer has access to app.MapManager. 

Also, the __esri namespace is only available starting with the arcgis-js-api 4 and above. Many of us are still using the 3.* versions of their api, which don't have that global defined.

**Comment by andy-ms** [comment]
Until there is some solution to Microsoft/TypeScript#14853, TypeScript doesn't have good support for your preferred style of dynamically importing from a library inside of a callback, instead of using a top-level import. If the `arcgis-js-api` package really is a collection of modules, this isn't the declaration file's fault; it shouldn't declare globals when the library isn't supposed to be global.

One option for you until Microsoft/TypeScript#14853 is fixed is to write an `adapter.d.ts` declaration file which manually imports types and redeclares them as globals.

```ts
import EsriMap = require("esri/map");

declare global {
  namespace esri {
    type Map = EsriMap;
  }
}
```

The other option is to use a module bundler and ES6 imports.

**Comment by TwitchBronBron** [comment]
I guess what I'm hearing from all of this is that you don't think that the namespace-style type definitions that I created belong in DefinitelyTyped?

**Comment by andy-ms** [comment]
I definitely don't want to duplicate the definition of a package to allow for global use of modules that aren't actually global. A real fix would be Microsoft/TypeScript#14853.
